<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Extract, QC and screen peptide epitopes â€¢ pepitope</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Extract, QC and screen peptide epitopes">
<meta name="description" content="Given a reference genome and VCF file, this package will provide the upstream/downstream peptide context of a variant. It will generate a summary report for protein-coding variants including the reference and mutated allele, read coverage, amino acid sequence, and other information. It can also be used to remove restriction sites from cDNA, alongside other helper functions.">
<meta property="og:description" content="Given a reference genome and VCF file, this package will provide the upstream/downstream peptide context of a variant. It will generate a summary report for protein-coding variants including the reference and mutated allele, read coverage, amino acid sequence, and other information. It can also be used to remove restriction sites from cDNA, alongside other helper functions.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">pepitope</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.990</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/variant.html">Variant calling</a></li>
<li class="nav-item"><a class="nav-link" href="articles/qc.html">Quality control</a></li>
<li class="nav-item"><a class="nav-link" href="articles/screen.html">Co-culture screen</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mschubert/pepitope/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="pepitope-extract-qc-and-screen-peptide-epitopes">pepitope: extract, qc and screen <em>pep</em>tide ep<em>itope</em>s<a class="anchor" aria-label="anchor" href="#pepitope-extract-qc-and-screen-peptide-epitopes"></a>
</h1></div>
<p>This R package is used to:</p>
<ul>
<li>extract peptides with flanking region around mutations (<a href="https://mschubert.github.io/pepitope/articles/variant.html" class="external-link">vignette</a>)</li>
<li>run quality control on sequencing of these libraries (<a href="https://mschubert.github.io/pepitope/articles/qc.html" class="external-link">vignette</a>)</li>
<li>perform differential abundance testing of co-culture screens (<a href="https://mschubert.github.io/pepitope/articles/screen.html" class="external-link">vignette</a>)</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>We require <code>R&gt;=4.5.0</code>, as this includes important fixes on Bioconductor.</p>
<p>The package is currently only available on Github, use the<code>remotes</code> package to install:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mschubert/pepitope"</span><span class="op">)</span></span></code></pre></div>
<p>In addition, we need the <a href="https://www.rust-lang.org/tools/install" class="external-link">Rust</a> programming language and its <code>cargo</code> package manager, which we use to install <a href="https://github.com/fulcrumgenomics/fqtk" class="external-link"><code>fqtk</code></a> and <a href="https://github.com/fulcrumgenomics/guide-counter" class="external-link"><code>guide-counter</code></a>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">cargo</span> install fqtk guide-counter</span></code></pre></div>
<p>We can check if the installation works by running, in R:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mschubert/pepitope" class="external-link">pepitope</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.which.html" class="external-link">Sys.which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fqtk"</span>, <span class="st">"guide-counter"</span><span class="op">)</span><span class="op">)</span> <span class="co"># should print paths, not ""</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level4">
<h4 id="generating-peptide-constructs-vignette">Generating peptide constructs (<a href="https://mschubert.github.io/pepitope/articles/variant.html" class="external-link">vignette</a>)<a class="anchor" aria-label="anchor" href="#generating-peptide-constructs-vignette"></a>
</h4>
<ul>
<li>Load a genome and annotation, usually GRCh38 and Ensembl</li>
<li>Load a VCF variants file as <code>VRanges</code> object</li>
<li>Annotate the protein-coding mutations</li>
<li>Optionally, loading a fusion VCF and annotating those</li>
<li>Subsetting the peptide context around each mutation</li>
<li>Make a report of variants, coding changes, and tiled peptides</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mschubert/pepitope" class="external-link">pepitope</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># genome and annotation</span></span>
<span><span class="va">ens106</span> <span class="op">=</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="st">"AH100643"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">asm</span> <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg38</a></span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">ens106</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span></span>
<span><span class="co"># read variants from VCF file, apply filters and annotate</span></span>
<span><span class="va">variant_vcf_file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"my_variants.vcf"</span>, package<span class="op">=</span><span class="st">"pepitope"</span><span class="op">)</span></span>
<span><span class="va">vr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">variant_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter_variants.html">filter_variants</a></span><span class="op">(</span>min_cov<span class="op">=</span><span class="fl">2</span>, min_af<span class="op">=</span><span class="fl">0.05</span>, pass<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ann</span> <span class="op">=</span> <span class="fu"><a href="reference/annotate_coding.html">annotate_coding</a></span><span class="op">(</span><span class="va">vr</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span></span>
<span><span class="va">subs</span> <span class="op">=</span> <span class="va">ann</span> <span class="op">|&gt;</span></span>
<span><span class="co">#    filter_expressed(rna_sample, min_reads=1, min_tpm=0) |&gt;</span></span>
<span>    <span class="fu"><a href="reference/subset_context.html">subset_context</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read fusion variants, apply filters and annotate</span></span>
<span><span class="va">fusion_vcf_file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"my_fusions.vcf"</span>, package<span class="op">=</span><span class="st">"pepitope"</span><span class="op">)</span></span>
<span><span class="va">vr2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">fusion_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter_fusions.html">filter_fusions</a></span><span class="op">(</span>min_reads<span class="op">=</span><span class="fl">2</span>, min_split_reads<span class="op">=</span><span class="fl">1</span>, min_tools<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">vr2</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span></span>
<span><span class="va">fus</span> <span class="op">=</span> <span class="fu"><a href="reference/annotate_fusions.html">annotate_fusions</a></span><span class="op">(</span><span class="va">vr2</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/subset_context_fusion.html">subset_context_fusion</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create construct tables and make a report</span></span>
<span><span class="va">tiled</span> <span class="op">=</span> <span class="fu"><a href="reference/make_peptides.html">make_peptides</a></span><span class="op">(</span><span class="va">subs</span>, <span class="va">fus</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/pep_tile.html">pep_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/remove_cutsite.html">remove_cutsite</a></span><span class="op">(</span>BbsI<span class="op">=</span><span class="st">"GAAGAC"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">report</span> <span class="op">=</span> <span class="fu"><a href="reference/make_report.html">make_report</a></span><span class="op">(</span><span class="va">ann</span>, <span class="va">subs</span>, <span class="va">fus</span>, <span class="va">tiled</span><span class="op">)</span></span>
<span><span class="fu">writexl</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/writexl//reference/write_xlsx.html" class="external-link">write_xlsx</a></span><span class="op">(</span><span class="va">report</span>, <span class="st">"my_variants.xlsx"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="creating-construct-library-wetlab">Creating construct library (wetlab)<a class="anchor" aria-label="anchor" href="#creating-construct-library-wetlab"></a>
</h4>
<ul>
<li>Decide on barcodes for each of those constructs</li>
<li>Add them in the annotation sheets as <code>barcode</code> or <code>barcode_{1,2}</code> etc.</li>
<li>Order these constructs as gene blocks and clone them into expression vectors</li>
<li>Transduce target cells with this peptide construct library</li>
</ul>
<p>Normally, we want to create an <code>xlsx</code> document with all libraries in use:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this file is manually created from the output of step 1</span></span>
<span><span class="va">fname</span> <span class="op">=</span> <span class="st">"my_combined_barcoded_file.xlsx"</span></span>
<span><span class="va">sheets</span> <span class="op">=</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/excel_sheets.html" class="external-link">excel_sheets</a></span><span class="op">(</span><span class="va">fname</span><span class="op">)</span></span>
<span><span class="va">all_constructs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sheets</span>, <span class="fu">readxl</span><span class="fu">::</span><span class="va"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span>, path<span class="op">=</span><span class="va">fname</span>, simplify<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Here, we use our example library instead:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># creating barcoded constructs</span></span>
<span><span class="va">lib</span> <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/hawkjo/freebarcodes/master/barcodes/barcodes12-1.txt"</span></span>
<span><span class="va">valid_barcodes</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">lib</span>, col_names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">$</span><span class="va">X1</span></span>
<span><span class="va">all_constructs</span> <span class="op">=</span> <span class="fu"><a href="reference/example_peptides.html">example_peptides</a></span><span class="op">(</span><span class="va">valid_barcodes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="performing-quality-control-on-construct-library-sequencing-vignette">Performing quality control on construct library sequencing (<a href="https://mschubert.github.io/pepitope/articles/qc.html" class="external-link">vignette</a>)<a class="anchor" aria-label="anchor" href="#performing-quality-control-on-construct-library-sequencing-vignette"></a>
</h4>
<ul>
<li>Check the quality of the construct libraries</li>
<li>Check the quality of the transduced target cells</li>
<li>Check the quality of the co-culture screens</li>
</ul>
<p>Quality Control can be performed on the library, the target cells, or the co-culture screen:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># demultiplexing and counting example data</span></span>
<span><span class="va">sample_sheet</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"my_samples.tsv"</span>, package<span class="op">=</span><span class="st">"pepitope"</span><span class="op">)</span></span>
<span><span class="va">fastq_file</span> <span class="op">=</span> <span class="fu"><a href="reference/example_fastq.html">example_fastq</a></span><span class="op">(</span><span class="va">sample_sheet</span>, <span class="va">all_constructs</span><span class="op">)</span></span>
<span><span class="va">temp_dir</span> <span class="op">=</span> <span class="fu"><a href="reference/demux_fq.html">demux_fq</a></span><span class="op">(</span><span class="va">fastq_file</span>, <span class="va">sample_sheet</span>, read_structures<span class="op">=</span><span class="st">"7B+T"</span><span class="op">)</span></span>
<span><span class="va">dset</span> <span class="op">=</span> <span class="fu"><a href="reference/count_bc.html">count_bc</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="va">all_constructs</span>, <span class="va">valid_barcodes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># quality control plots</span></span>
<span><span class="fu"><a href="reference/plot_barcode_overlap.html">plot_barcode_overlap</a></span><span class="op">(</span><span class="va">all_constructs</span>, <span class="va">valid_barcodes</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_reads.html">plot_reads</a></span><span class="op">(</span><span class="va">dset</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_distr.html">plot_distr</a></span><span class="op">(</span><span class="va">dset</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="differential-abundance-testing-of-co-culture-screens-vignette">Differential abundance testing of co-culture screens (<a href="https://mschubert.github.io/pepitope/articles/screen.html" class="external-link">vignette</a>)<a class="anchor" aria-label="anchor" href="#differential-abundance-testing-of-co-culture-screens-vignette"></a>
</h4>
<p>Differential abundance testing works the following way:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform abundance testing and plot results</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/screen_calc.html">screen_calc</a></span><span class="op">(</span><span class="va">dset</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"Mock"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_screen.html">plot_screen</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">`Sample vs Mock`</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/mschubert/pepitope/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/mschubert/pepitope/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pepitope</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Michael Schubert <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6862-5221" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Schubert.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
