---
title: "Quality control"
output:
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Quality control}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{css echo=FALSE}
img {
    border: 0px !important;
    margin: 2em 2em 2em 2em !important;
}
code {
    border: 0px !important;
}
```

```{r echo=FALSE, results="hide"}
knitr::opts_chunk$set(
    cache = FALSE,
    echo = TRUE,
    collapse = TRUE,
    comment = "#>"
)
suppressPackageStartupMessages(library(pepitope))
```

Prerequisites
-------------

This guide shows how to perform quality control on a library using the barcoded
constructs from Variant Calling and a sample sheet that contains a second,
sample-level barcode.

The library itself is generated by ordering the barcoded constructs and then
cloning them into the appropriate expression vectors to function as
MHC-presented peptides.

This workflow can be used either on the construct library, the transduced
cells, or cells after co-culture. The goal of this step is to confirm that all
constructs are present in the approriate proportions. We provide functions to
process and analyze the sequencing data.

For this step, we will work with a single FASTQ file that can contain multiple
samples:

```{r}
fastq = ""
```

Sample demultiplexing
---------------------

### Creating a sample sheet

```{r}
samples = 
```

__ SHOW TABLE HERE __

### Demultiplexing

This step is using the `fqtk` toolkit to split the multi-sample FASTQ file into
individual sample files, separated by their sample barcode and save to a
temporary directory.

Here, `read_structures` describes how the split the samples. It has the
following possible fields, preceded by the number of nucleotides in the read. A
`+` instead of a number is used to indicated to use all remaining nucleotides:

* `B` -- the sample barcode to split on
* `T` -- the read sequence with the construct barcode and sequence
* `S` -- skip these nucleotides and do not include in output

In this example, we are using a 7-nucleotide barcode for the samples and keep
the rest of the read:

```{r}
temp_dir = demux_fq(fastq, samples, read_structures="7B+T")
```

You can perform additional quality control steps on the sample-level FASTQ
files, e.g. using `fastqc` or `multiqc` (not included in this package).

Construct barcode counting
--------------------------

### Working with the full construct barcode library

```{r}
barcode_lib = "/path/to/file"
```

### Counting construct barcodes

The next step is to count the construct barcodes in each individual FASTQ file
after demultiplexing. Internally, we use the `guide-counter` tool to identify the
position the construct barcodes occur in the library reads and subsequently
count the number of the occurrences:

```{r}
counts = count_bc(temp_dir, barcode_lib) #FIXME: load_dset
```

Here, `counts` will be a list with two objects, `meta` and `counts`:

* `meta` -- is a data.frame sample information in rows, including total and mapped reads
* `counts` -- is a matrix with constructs in rows and samples in columns

### Merging with construct info

merge with construct info

Quality Control plots
---------------------

### Plotting total barcode counts

static + interactive plot

### Representation of individual barcodes

static + interactive plot

### Correlation of read counts between samples

static plot

