<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quality control • pepitope</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quality control">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">pepitope</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.917</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/variant.html">Variant calling</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/qc.html">Quality control</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/screen.html">Co-culture screen</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mschubert/pepitope/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quality control</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mschubert/pepitope/blob/HEAD/vignettes/qc.Rmd" class="external-link"><code>vignettes/qc.Rmd</code></a></small>
      <div class="d-none name"><code>qc.Rmd</code></div>
    </div>

    
    
<style type="text/css">
img {
    border: 0px !important;
    margin: 2em 2em 2em 2em !important;
}
code {
    border: 0px !important;
}
</style>
<p>This guide shows how to perform quality control on a library using
the barcoded constructs from Variant Calling and a sample sheet that
contains a second, sample-level barcode.</p>
<p>The library itself is generated by ordering the barcoded constructs
and then cloning them into the appropriate expression vectors to
function as MHC-presented peptides.</p>
<p>This workflow can be used either on the construct library, the
transduced cells, or cells after co-culture. The goal of this step is to
confirm that all constructs are present in the approriate proportions.
We provide functions to process and analyze the sequencing data.</p>
<div class="section level2">
<h2 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h2>
<div class="section level3">
<h3 id="creating-a-sample-sheet">Creating a sample sheet<a class="anchor" aria-label="anchor" href="#creating-a-sample-sheet"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">=</span> <span class="st">""</span></span></code></pre></div>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="sequencing-output">Sequencing output<a class="anchor" aria-label="anchor" href="#sequencing-output"></a>
</h3>
<p>For this workflow, we need with a single FASTQ file that can contain
multiple samples:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastq</span> <span class="op">=</span> <span class="st">""</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="sample-demultiplexing">Sample demultiplexing<a class="anchor" aria-label="anchor" href="#sample-demultiplexing"></a>
</h2>
<div class="section level3">
<h3 id="feature-syntax">Feature syntax<a class="anchor" aria-label="anchor" href="#feature-syntax"></a>
</h3>
<p>This step is using the <code>fqtk</code> toolkit to split the
multi-sample FASTQ file into individual sample files, separated by their
sample barcode and save to a temporary directory.</p>
<p>Here, <code>read_structures</code> describes how the split the
samples. It has the following possible fields, preceded by the number of
nucleotides in the read. A <code>+</code> instead of a number is used to
indicated to use all remaining nucleotides:</p>
<ul>
<li>
<code>B</code> – the sample barcode to split on</li>
<li>
<code>T</code> – the read sequence with the construct barcode and
sequence</li>
<li>
<code>S</code> – skip these nucleotides and do not include in
output</li>
</ul>
</div>
<div class="section level3">
<h3 id="demultiplexing-call">Demultiplexing call<a class="anchor" aria-label="anchor" href="#demultiplexing-call"></a>
</h3>
<p>In this example, we are using a 7-nucleotide barcode for the samples
and keep the rest of the read:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">=</span> <span class="fu">demux_fq</span><span class="op">(</span><span class="va">fastq</span>, <span class="va">samples</span>, read_structures<span class="op">=</span><span class="st">"7B+T"</span><span class="op">)</span></span></code></pre></div>
<p>You can perform additional quality control steps on the sample-level
FASTQ files, e.g. using <code>fastqc</code> or <code>multiqc</code> (not
included in this package).</p>
</div>
</div>
<div class="section level2">
<h2 id="construct-barcode-counting">Construct barcode counting<a class="anchor" aria-label="anchor" href="#construct-barcode-counting"></a>
</h2>
<div class="section level3">
<h3 id="working-with-the-full-construct-barcode-library">Working with the full construct barcode library<a class="anchor" aria-label="anchor" href="#working-with-the-full-construct-barcode-library"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barcode_lib</span> <span class="op">=</span> <span class="st">"/path/to/file"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="counting-construct-barcodes">Counting construct barcodes<a class="anchor" aria-label="anchor" href="#counting-construct-barcodes"></a>
</h3>
<p>The next step is to count the construct barcodes in each individual
FASTQ file after demultiplexing. Internally, we use the
<code>guide-counter</code> tool to identify the position the construct
barcodes occur in the library reads and subsequently count the number of
occurrences. Here we pass the directory containing the demultiplexed
FASTQ files and the barcode library file with all possible (not only the
used) barcodes:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dset</span> <span class="op">=</span> <span class="fu"><a href="../reference/count_bc.html">count_bc</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="va">barcode_lib</span><span class="op">)</span></span></code></pre></div>
<p>Here, <code>dset</code> will be a list with two objects:</p>
<ul>
<li>
<code>meta</code> – is a data.frame sample and read information in
rows</li>
<li>
<code>counts</code> – is a matrix with constructs in rows and
samples in columns</li>
</ul>
</div>
<div class="section level3">
<h3 id="merging-with-construct-info">Merging with construct info<a class="anchor" aria-label="anchor" href="#merging-with-construct-info"></a>
</h3>
<p>In this step, we add the information of which barcode belongs to
which sample and construct. We can use multiple barcode annotation
files, but it is required that every construct barcodes occurs only in
one sample and construct:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcs</span> <span class="op">=</span> <span class="st">""</span> <span class="co"># load construct info files here</span></span>
<span><span class="va">reps</span> <span class="op">=</span> <span class="fu">calc_representation</span><span class="op">(</span><span class="va">dset</span><span class="op">$</span><span class="va">counts</span>, <span class="va">bcs</span>, <span class="va">dset</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quality-control-plots">Quality Control plots<a class="anchor" aria-label="anchor" href="#quality-control-plots"></a>
</h2>
<div class="section level3">
<h3 id="plotting-total-barcode-counts">Plotting total barcode counts<a class="anchor" aria-label="anchor" href="#plotting-total-barcode-counts"></a>
</h3>
<p>The first overview that we want to get is to know how many barcodes
are in which demultiplexed FASTQ, and whether they match the sample we
expect them to be from. We can plot this the following way, for total
read counts on the left and number of barcodes that have 10 or more
reads on the right:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_reads</span><span class="op">(</span><span class="va">reps</span>, <span class="va">dset</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
<p>We can also plot this interactively with <code>plotly</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="va">plot</span> <span class="op">=</span> <span class="fu">plot_distr</span><span class="op">(</span><span class="va">reps</span><span class="op">)</span></span>
<span><span class="fu">subplot</span><span class="op">(</span><span class="fu">ggplotly</span><span class="op">(</span><span class="va">plot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu">ggplotly</span><span class="op">(</span><span class="va">plot</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, nrows<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="representation-of-individual-barcodes">Representation of individual barcodes<a class="anchor" aria-label="anchor" href="#representation-of-individual-barcodes"></a>
</h3>
<p>The next question we want to ask is whether the individual construct
barcodes are equally distributed within a sample, and where any
potential contaminations come from. For this, we order the barcodes from
least abundant (left) to most abundant (right) and plot a continuous
line for how many reads are sequenced of this barcode on the y axis:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_distr</span><span class="op">(</span><span class="va">reps</span><span class="op">)</span></span></code></pre></div>
<p>In a perflectly balanced library we would expect a horizonal line
that stays constant. This, however, is impossible to reach due to the
sampling process that obtains reads stochastically. As the above plot is
sampled from a near-perfect library, this distribution is the best you
can expect.</p>
<p>We can also plot this interactively with <code>plotly</code>, where
we can hover over with the mouse to see which barcode is in which
position exactly:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="fu">ggplotly</span><span class="op">(</span><span class="fu">plot_distr</span><span class="op">(</span><span class="va">reps</span><span class="op">)</span>, height<span class="op">=</span><span class="fl">800</span><span class="op">)</span></span></code></pre></div>
<!-- add cor plots here
### Correlation of read counts between samples

static plot


-->
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Schubert.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
