<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Variant calling • pepitope</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variant calling">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">pepitope</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.917</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/variant.html">Variant calling</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/qc.html">Quality control</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/screen.html">Co-culture screen</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mschubert/pepitope"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Variant calling</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mschubert/pepitope/blob/HEAD/vignettes/variant.Rmd" class="external-link"><code>vignettes/variant.Rmd</code></a></small>
      <div class="d-none name"><code>variant.Rmd</code></div>
    </div>

    
    
<style type="text/css">
img {
    border: 0px !important;
    margin: 2em 2em 2em 2em !important;
}
code {
    border: 0px !important;
}
</style>
<p>This vignette describes how to go from variants (and optionally gene
fusions) to the Minigene Library report.</p>
<p>We will usually work with standard processing pipelines for variant
calling, RNA-seq quantification, and RNA fusion calling. The best-known
collection of these workflows is <a href="https://nf-co.re/" class="external-link">NF-core</a>.</p>
<p>It provides the following pipelines for these tasks:</p>
<ul>
<li>
<a href="https://nf-co.re/sarek/" class="external-link"><code>sarek</code></a> for variant
calling using HaploytypeCaller and Mutect2</li>
<li>
<a href="https://nf-co.re/rnaseq/" class="external-link"><code>rnaseq</code></a> for
RNA-seq quantification using STAR</li>
<li>
<a href="https://nf-co.re/rnafusion/" class="external-link"><code>rnafusion</code></a> for
RNA fusion quantification using multiple algorithms</li>
</ul>
<p>Running these will usually require access to high performance
computing facilities, and you already get the processed files from your
core facility or sequencing provider.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mschubert/pepitope" class="external-link">pepitope</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># SNPs and small indels, e.g. from 'sarek' nf-core pipeline</span></span>
<span><span class="va">variant_vcf_file</span> <span class="op">=</span> <span class="st">"my_variants.vcf.gz"</span></span>
<span></span>
<span><span class="co"># RNA counts and TPM file, e.g. from 'rnaseq' nf-core pipeline</span></span>
<span><span class="va">rna_counts_file</span> <span class="op">=</span> <span class="st">"count_file.tsv"</span></span>
<span><span class="va">rna_tpm_file</span> <span class="op">=</span> <span class="st">"tpm_file.tsv"</span></span>
<span></span>
<span><span class="co"># Combined fusion VCF file, e.g. from 'rnafusion' nf-core pipeline</span></span>
<span><span class="va">fusion_vcf_file</span> <span class="op">=</span> <span class="st">"my_fusions.vcf.gz"</span></span></code></pre></div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<div class="section level3">
<h3 id="selecting-the-right-reference-genome">Selecting the right reference genome<a class="anchor" aria-label="anchor" href="#selecting-the-right-reference-genome"></a>
</h3>
<p>The variant calling and RNA-seq counts were mapped to a reference
genome and gene annotations. It is important to keep these consistent
between the NF-core processing pipelines and the Minigene Library
annotation.</p>
<p>We are usually working with <code>BSgenome</code> (for the reference
genome) and <code>EnsDb</code> (for the gene annotations) objects. For
human data, the most widely used reference genome is
<code>GRCh38</code>, and a recent <code>Ensembl</code> annotation
release.</p>
<p>With our test data, we know that <code>GRCh38</code> is the correct
reference genome and <code>Ensembl 106</code> is the correct version of
gene annotations. We can get both objects from the <code>BSgenome</code>
and <code>AnnotationHub</code> Bioconductor packages, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens106</span> <span class="op">=</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="st">"AH100643"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">asm</span> <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.NCBI.GRCh38/man/package.html" class="external-link">BSgenome.Hsapiens.NCBI.GRCh38</a></span></span>
<span><span class="co"># rna_counts = readr::read_tsv(rna_counts_file)</span></span></code></pre></div>
<p>A caveat here is that the chromosome prefixes need to be consistent
between the variants in the VCF file and the genome/gene annotations.
There are two “styles”, either UCSC (includes “chr” prefix) or
NCBI/Ensembl (without “chr” prefix).</p>
<p>The <code>sarek</code> pipline uses UCSC prefixes on the GRCh38
genome, so we need to switch the genome and gene annotation styles:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">ens106</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">asm</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span><span class="va">asm</span><span class="op">@</span><span class="va">seqinfo</span><span class="op">@</span><span class="va">genome</span><span class="op">[</span><span class="op">]</span> <span class="op">=</span> <span class="st">"GRCh38"</span></span></code></pre></div>
<p>The correct styles will depend on how your VCF files were
generated.</p>
</div>
<div class="section level3">
<h3 id="adding-rna-expression">Adding RNA expression<a class="anchor" aria-label="anchor" href="#adding-rna-expression"></a>
</h3>
<p>The NF-core <code>rnaseq</code> workflow will provide two gene
expression files, one for raw read counts and one for transcripts per
million (TPM). These contain all samples in a run, so we need to subset
them to the current sample we are interested in. These files are usually
called:</p>
<ul>
<li><code>salmon.merged.gene_counts.tsv</code></li>
<li><code>salmon.merged.gene_tpm.tsv</code></li>
</ul>
<p>We can combine and subset them the following way:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_counts</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">rna_counts_file</span><span class="op">)</span></span>
<span><span class="va">rna_tpm</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">rna_tpm_file</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">args</span><span class="op">$</span><span class="va">rna_count</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span>, count<span class="op">=</span><span class="va">SAMPLE</span><span class="op">)</span></span>
<span><span class="va">tpm</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">args</span><span class="op">$</span><span class="va">rna_tpm</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span>, tpm<span class="op">=</span><span class="va">SAMPLE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna_sample</span> <span class="op">=</span> <span class="fu">inner_join</span><span class="op">(</span><span class="va">counts</span>, <span class="va">tpm</span><span class="op">)</span></span></code></pre></div>
<p>__ SHOW TABLE HERE __</p>
</div>
</div>
<div class="section level2">
<h2 id="snps-and-small-indels">SNPs and small indels<a class="anchor" aria-label="anchor" href="#snps-and-small-indels"></a>
</h2>
<div class="section level3">
<h3 id="reading-and-filtering-mutations">Reading and filtering mutations<a class="anchor" aria-label="anchor" href="#reading-and-filtering-mutations"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vr1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">variant_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_variants.html">filter_variants</a></span><span class="op">(</span>min_cov<span class="op">=</span><span class="fl">2</span>, min_af<span class="op">=</span><span class="fl">0.05</span>, pass<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_cov=2</code> – a variant needs to be covered by at least 2
reads</li>
<li>
<code>min_af=0.05</code> – a variant needs to occur in at leat 5% of
reads</li>
<li>
<code>pass=TRUE</code> – a variant needs to pass the standard QC
filters</li>
</ul>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="annotating-and-subsetting-expressed-variants">Annotating and subsetting expressed variants<a class="anchor" aria-label="anchor" href="#annotating-and-subsetting-expressed-variants"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ann</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_coding.html">annotate_coding</a></span><span class="op">(</span><span class="va">vr1</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span></span>
<span><span class="va">subs</span> <span class="op">=</span> <span class="va">ann</span> <span class="op">|&gt;</span></span>
<span><span class="co">#    filter_expressed(rna_sample, min_reads=1, min_tpm=0) |&gt;</span></span>
<span>    <span class="fu"><a href="../reference/subset_context.html">subset_context</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_reads=1</code> – the gene needs to have at least one RNA
read</li>
<li>
<code>min_tpm=0</code> – we do not apply an additional TPM
filter</li>
</ul>
<p>In addition, we set the region of interest (context) to 15 codons up-
and downstream of the variant. Hence, a SNP will have a total length of
93 nucleotides (15*3 + the SNP codon itself). An insertion will have the
inserted sequence and 15 codons, a deletion only 15 codons both sides. A
frameshift will have 15 codons upstream and the entire sequence
downstream until a STOP codon is reached. The latter may extend into the
3’ UTR.</p>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="phasing-with-germline-variants">Phasing with germline variants<a class="anchor" aria-label="anchor" href="#phasing-with-germline-variants"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># not implemented yet</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="fusion-genes-from-rna-seq">Fusion genes from RNA-seq<a class="anchor" aria-label="anchor" href="#fusion-genes-from-rna-seq"></a>
</h2>
<div class="section level3">
<h3 id="annotating-fusion-genes">Annotating fusion genes<a class="anchor" aria-label="anchor" href="#annotating-fusion-genes"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vr2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">fusion_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_fusions.html">filter_fusions</a></span><span class="op">(</span>min_reads<span class="op">=</span><span class="fl">2</span>, min_split_reads<span class="op">=</span><span class="fl">1</span>, min_tools<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">vr2</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_reads=2</code> – the fusion needs to be supported by 2
split or pair distance reads</li>
<li>
<code>min_split_reads=1</code> – the fusion needs to be supported by
at least one split read</li>
<li>
<code>min_tools=1</code> – the fusion needs to be reported by at
least one tool</li>
</ul>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="subsetting-expressed-fusion-genes">Subsetting expressed fusion genes<a class="anchor" aria-label="anchor" href="#subsetting-expressed-fusion-genes"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fus</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_fusions.html">annotate_fusions</a></span><span class="op">(</span><span class="va">vr2</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">filter_expressed</span><span class="op">(</span><span class="va">rna_sample</span>, min_reads<span class="op">=</span><span class="fl">1</span>, min_tpm<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/subset_context.html">subset_context</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_reads=1</code> – the gene needs to have at least one RNA
read</li>
<li>
<code>min_tpm=0</code> – we do not apply an additional TPM
filter</li>
</ul>
<p>In addition, we set the region of interest to 15 codons up- and
downstream of the fusion site.</p>
<p>__ SHOW TABLE HERE __</p>
</div>
</div>
<div class="section level2">
<h2 id="generating-the-minigene-library">Generating the Minigene Library<a class="anchor" aria-label="anchor" href="#generating-the-minigene-library"></a>
</h2>
<div class="section level3">
<h3 id="tiling-cdnas-of-interest-into-smaller-peptides">Tiling cDNAs of interest into smaller peptides<a class="anchor" aria-label="anchor" href="#tiling-cdnas-of-interest-into-smaller-peptides"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tiled</span> <span class="op">=</span> <span class="fu"><a href="../reference/pep_tile.html">pep_tile</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/remove_cutsite.html">remove_cutsite</a></span><span class="op">(</span>BbsI<span class="op">=</span><span class="st">"GAAGAC"</span><span class="op">)</span></span></code></pre></div>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="adding-barcodes">Adding barcodes<a class="anchor" aria-label="anchor" href="#adding-barcodes"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is not yet implemented</span></span></code></pre></div>
<p>__ SHOW TABLE HERE __</p>
</div>
<div class="section level3">
<h3 id="saving-a-report-file">Saving a report file<a class="anchor" aria-label="anchor" href="#saving-a-report-file"></a>
</h3>
<p>We can then combine our generated tables into a report save it with
the <code>writexl</code> package:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report</span> <span class="op">=</span> <span class="fu"><a href="../reference/make_report.html">make_report</a></span><span class="op">(</span><span class="va">ann</span>, <span class="va">subs</span>, <span class="va">fus</span>, <span class="va">tiled</span><span class="op">)</span></span>
<span><span class="fu">writexl</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/writexl//reference/write_xlsx.html" class="external-link">write_xlsx</a></span><span class="op">(</span><span class="va">report</span>, <span class="st">"report_file.xlsx"</span><span class="op">)</span></span></code></pre></div>
<p>This <code>.xlsx</code> file will contain the different tables as
sheets. We will use it as an annotation file in the quality control and
screen steps.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Schubert.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
