<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Variant calling • pepitope</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variant calling">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">pepitope</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.919</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/variant.html">Variant calling</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/qc.html">Quality control</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/screen.html">Co-culture screen</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mschubert/pepitope/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="variant_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="variant_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="variant_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="variant_files/datatables-binding-0.33/datatables.js"></script><link href="variant_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="variant_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="variant_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="variant_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="variant_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Variant calling</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mschubert/pepitope/blob/HEAD/vignettes/variant.Rmd" class="external-link"><code>vignettes/variant.Rmd</code></a></small>
      <div class="d-none name"><code>variant.Rmd</code></div>
    </div>

    
    
<style type="text/css">
img {
    border: 0px !important;
    margin: 2em 2em 2em 2em !important;
}
code {
    border: 0px !important;
}
table.dataTable.compact,
.dataTables_wrapper .dataTables_paginate,
.dataTables_wrapper .dataTables_info,
.dataTables_wrapper .dataTables_filter,
.dataTables_wrapper .dataTables_length {
    font-size: 0.7em;
}
</style>
<p>This vignette describes how to go from variants (and optionally gene
fusions) to the Minigene Library report.</p>
<p>We will usually work with standard processing pipelines for variant
calling, RNA-seq quantification, and RNA fusion calling. The best-known
collection of these workflows is <a href="https://nf-co.re/" class="external-link">NF-core</a>.</p>
<p>It provides the following pipelines for these tasks:</p>
<ul>
<li>
<a href="https://nf-co.re/sarek/" class="external-link"><code>sarek</code></a> for variant
calling using HaploytypeCaller and Mutect2</li>
<li>
<a href="https://nf-co.re/rnaseq/" class="external-link"><code>rnaseq</code></a> for
RNA-seq quantification using STAR and Salmon</li>
<li>
<a href="https://nf-co.re/rnafusion/" class="external-link"><code>rnafusion</code></a> for
RNA fusion quantification using multiple algorithms</li>
</ul>
<p>Running these will usually require access to high performance
computing facilities, and you already get the processed files from your
core facility or sequencing provider.</p>
<p>Here, we are using example files provided with the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mschubert/pepitope" class="external-link">pepitope</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># SNPs and small indels, e.g. from 'sarek' nf-core pipeline</span></span>
<span><span class="va">variant_vcf_file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"my_variants.vcf"</span>, package<span class="op">=</span><span class="st">"pepitope"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined fusion VCF file, e.g. from 'rnafusion' nf-core pipeline</span></span>
<span><span class="va">fusion_vcf_file</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"my_fusions.vcf"</span>, package<span class="op">=</span><span class="st">"pepitope"</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h2>
<div class="section level3">
<h3 id="selecting-the-right-reference-genome">Selecting the right reference genome<a class="anchor" aria-label="anchor" href="#selecting-the-right-reference-genome"></a>
</h3>
<p>The variant calling and RNA-seq counts were mapped to a reference
genome and gene annotations. It is important to keep these consistent
between the NF-core processing pipelines and the Minigene Library
annotation.</p>
<p>We are usually working with <code>BSgenome</code> (for the reference
genome) and <code>EnsDb</code> (for the gene annotations) objects. For
human data, the most widely used reference genome is
<code>GRCh38</code>, and a recent <code>Ensembl</code> annotation
release.</p>
<p>With our test data, we know that <code>GRCh38</code> is the correct
reference genome and <code>Ensembl 106</code> is the correct version of
gene annotations. We can get both objects from the <code>BSgenome</code>
and <code>AnnotationHub</code> Bioconductor packages, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens106</span> <span class="op">=</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="st">"AH100643"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; require("ensembldb")</span></span>
<span><span class="va">asm</span> <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.NCBI.GRCh38/man/package.html" class="external-link">BSgenome.Hsapiens.NCBI.GRCh38</a></span></span></code></pre></div>
<p>A caveat here is that the chromosome prefixes need to be consistent
between the variants in the VCF file and the genome/gene annotations.
There are two “styles”, either UCSC (includes “chr” prefix) or
NCBI/Ensembl (without “chr” prefix).</p>
<p>The <code>sarek</code> pipline uses UCSC prefixes on the GRCh38
genome, so we need to switch the genome and gene annotation styles:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">ens106</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">asm</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span><span class="va">asm</span><span class="op">@</span><span class="va">seqinfo</span><span class="op">@</span><span class="va">genome</span><span class="op">[</span><span class="op">]</span> <span class="op">=</span> <span class="st">"GRCh38"</span></span></code></pre></div>
<p>The correct styles will depend on how your VCF files were
generated.</p>
</div>
<div class="section level3">
<h3 id="adding-rna-expression">Adding RNA expression<a class="anchor" aria-label="anchor" href="#adding-rna-expression"></a>
</h3>
<p>The <a href="https://nf-co.re/rnaseq/" class="external-link">NF-core <code>rnaseq</code>
workflow</a> will provide two gene expression files, one for raw read
counts and one for transcripts per million (TPM). These contain all
samples in a run, so we need to subset them to the current sample we are
interested in. These files are usually called:</p>
<ul>
<li><code>salmon.merged.gene_counts.tsv</code></li>
<li><code>salmon.merged.gene_tpm.tsv</code></li>
</ul>
<p>We can combine and subset them the following way:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note that this is not run in this example because we don't have RNA-seq data</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"salmon.merged.gene_counts.tsv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span>, count<span class="op">=</span><span class="va">SAMPLE</span><span class="op">)</span></span>
<span><span class="va">tpm</span> <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"salmon.merged.gene_tpm.tsv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span>, tpm<span class="op">=</span><span class="va">SAMPLE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna_sample</span> <span class="op">=</span> <span class="fu">inner_join</span><span class="op">(</span><span class="va">counts</span>, <span class="va">tpm</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="snps-and-small-indels">SNPs and small indels<a class="anchor" aria-label="anchor" href="#snps-and-small-indels"></a>
</h2>
<div class="section level3">
<h3 id="reading-and-filtering-mutations">Reading and filtering mutations<a class="anchor" aria-label="anchor" href="#reading-and-filtering-mutations"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vr1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">variant_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_variants.html">filter_variants</a></span><span class="op">(</span>min_cov<span class="op">=</span><span class="fl">2</span>, min_af<span class="op">=</span><span class="fl">0.05</span>, pass<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_cov=2</code> – a variant needs to be covered by at least 2
reads</li>
<li>
<code>min_af=0.05</code> – a variant needs to occur in at leat 5% of
reads</li>
<li>
<code>pass=TRUE</code> – a variant needs to pass the standard QC
filters</li>
</ul>
<p>The resulting <code>vr1</code> object looks like the following:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41"],["chr1","chr1","chr2","chr3","chr3","chr3","chr3","chr3","chr3","chr5","chr5","chr5","chr6","chr6","chr7","chr7","chr7","chr7","chr7","chr7","chr8","chr9","chr10","chr12","chr12","chr12","chr12","chr16","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr19","chr19","chr20"],[111494366,115256528,179612938,10183630,178936090,178936091,178936092,178952084,178952085,1295228,112175591,112175610,152419277,152419327,55242460,55242464,55242465,55249070,55249071,55259515,37650672,21970916,89623105,25398281,25398284,25398284,25398285,68862898,7574016,7577114,7577120,7577121,7577537,7577538,7578190,7578454,7578455,41242962,2307109,15349235,31022441],[111494366,115256528,179612938,10183630,178936092,178936091,178936092,178952084,178952085,1295228,112175591,112175610,152419277,152419327,55242460,55242464,55242465,55249071,55249071,55259515,37650672,21970916,89623105,25398281,25398284,25398284,25398285,68862898,7574016,7577114,7577120,7577121,7577538,7577538,7578190,7578454,7578455,41242962,2307109,15349235,31022441],[1,1,1,1,3,1,1,1,1,1,1,1,1,1,1,1,1,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,1,1,1,1,1,1,1,1],["*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*","*"],["C","G","C","G","CTT","C","A","A","A","C","G","A","C","G","G","T","T","TC","C","C","C","G","C","G","G","G","A","C","G","G","G","G","CG","C","G","G","T","A","C","C","T"],["T","A","T","A","C","A","AG","AT","G","T","A","G","T","A","GT","TAA","A","T","T","T","A","T","T","C","T","A","G","T","A","A","T","A","C","T","A","GA","TG","T","T","T","G"],[90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90],[10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10],["SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE"]],"container":"<table class=\"compact stripe hover\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>seqnames<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>refDepth<\/th>\n      <th>altDepth<\/th>\n      <th>sampleNames<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"seqnames","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"ref","targets":6},{"name":"alt","targets":7},{"name":"refDepth","targets":8},{"name":"altDepth","targets":9},{"name":"sampleNames","targets":10}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="annotating-and-subsetting-expressed-variants">Annotating and subsetting expressed variants<a class="anchor" aria-label="anchor" href="#annotating-and-subsetting-expressed-variants"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ann</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_coding.html">annotate_coding</a></span><span class="op">(</span><span class="va">vr1</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span></span>
<span><span class="va">subs</span> <span class="op">=</span> <span class="va">ann</span> <span class="op">|&gt;</span></span>
<span><span class="co">#    filter_expressed(rna_sample, min_reads=1, min_tpm=0) |&gt;</span></span>
<span>    <span class="fu"><a href="../reference/subset_context.html">subset_context</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_reads=1</code> – the gene needs to have at least one RNA
read</li>
<li>
<code>min_tpm=0</code> – we do not apply an additional TPM
filter</li>
</ul>
<p>In addition, we set the region of interest (context) to 15 codons up-
and downstream of the variant. Hence, a SNP will have a total length of
93 nucleotides (15*3 + the SNP codon itself). An insertion will have the
inserted sequence and 15 codons, a deletion only 15 codons both sides. A
frameshift will have 15 codons upstream and the entire sequence
downstream until a STOP codon is reached. The latter may extend into the
3’ UTR.</p>
<p>The <code>subs</code> dataframe looks like the following:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20"],["chr9","chr9","chr9","chr9","chr9","chr9","chr9","chr9","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17","chr17"],[21970916,21970916,21970916,21970916,21970916,21970916,21970916,21970916,7577114,7577114,7577114,7577120,7577120,7577120,7577121,7577121,7577121,7578190,7578454,7578455],[21970916,21970916,21970916,21970916,21970916,21970916,21970916,21970916,7577114,7577114,7577114,7577120,7577120,7577120,7577121,7577121,7577121,7578190,7578454,7578455],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["-","-","-","-","-","-","-","-","+","+","+","+","+","+","+","+","+","+","+","+"],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],["SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE","SAMPLE"],["G","G","G","G","G","G","G","G","G","G","G","G","G","G","G","G","G","G","G","T"],["T","T","T","T","T","T","T","T","A","A","A","T","T","T","A","A","A","A","GA","TG"],[90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90,90],[10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10],["A","A","A","A","A","A","A","A","A","A","A","T","T","T","A","A","A","A","GA","TG"],[443,290,290,290,443,290,290,443,573,573,573,579,579,579,580,580,580,1022,1189,1190],[443,290,290,290,443,290,290,443,573,573,573,579,579,579,580,580,580,1022,1189,1190],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],[[148],[97],[97],[97],[148],[97],[97],[148],[191],[191],[191],[193],[193],[193],[194],[194],[194],[341],[397],[397]],[22,22,22,22,22,22,22,22,30,30,30,31,31,31,32,32,32,35,36,37],["ENST00000304494","ENST00000479692","ENST00000494262","ENST00000497750","ENST00000498124","ENST00000498628","ENST00000578845","LRG_11t1","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000293831","ENST00000293831"],[[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]],["ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","LRG_11","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960"],["nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","synonymous","synonymous","synonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","nonsynonymous","frameshift","frameshift"],["GCG","GCG","GCG","GCG","GCG","GCG","GCG","GCG","GGA","GGA","GGA","AAG","AAG","AAG","GAC","GAC","GAC","GTT","GAA","GAA"],["GAG","GAG","GAG","GAG","GAG","GAG","GAG","GAG","GGA","GGA","GGA","AAT","AAT","AAT","AAC","AAC","AAC","GAT","GAAA","GTGA"],["A","A","A","A","A","A","A","A","G","G","G","K","K","K","D","D","D","V","",""],["E","E","E","E","E","E","E","E","G","G","G","N","N","N","N","N","N","D","",""],["CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1"],["ENST00000304494","ENST00000479692","ENST00000494262","ENST00000497750","ENST00000498124","ENST00000498628","ENST00000578845","LRG_11t1","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000577269","ENST00000582746","ENST00000293831","ENST00000293831","ENST00000293831"],["GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGGTGAGGAC","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGTTTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCTGACCTCATC","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCTGACCTCATC"],["VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","AAGGTRGSNHARIDAAEGPSVTASIQVPGGE","VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","DVARYLRAAAGGTRGSNHARIDAAEGPSGED","AAGGTRGSNHARIDAAEGPSEMIGNHLWVCR","VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","MFVLDEADEMLSRGFKDQIYDIFQKLNSNTQ","MFVLDEADEMLSRGFKDQIYDIFQKLNSNTQ","MFVLDEADEMLSRGFKDQIYDIFQKLNSNTQ","FVLDEADEMLSRGFKDQIYDIFQKLNSNTQV","FVLDEADEMLSRGFKDQIYDIFQKLNSNTQV","FVLDEADEMLSRGFKDQIYDIFQKLNSNTQV","MITTDLLARGIDVQQVSLVINYDLPTNRENY","INMVTEEDKRTLRDIETFYNTSIEEMPLNVADLI","INMVTEEDKRTLRDIETFYNTSIEEMPLNVADLI"],["GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGGTGAGGAC","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAATGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAATGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAATGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGAACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGAACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGAACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGATTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAAATGCCCCTCAATGTTGC","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGTGAATGCCCCTCAATGTTGC"],["VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","AAGGTRGSNHARIDAEEGPSVTASIQVPGGE","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","DVARYLRAAAGGTRGSNHARIDAEEGPSGED","AAGGTRGSNHARIDAEEGPSEMIGNHLWVCR","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","MFVLDEADEMLSRGFNDQIYDIFQKLNSNTQ","MFVLDEADEMLSRGFNDQIYDIFQKLNSNTQ","MFVLDEADEMLSRGFNDQIYDIFQKLNSNTQ","FVLDEADEMLSRGFKNQIYDIFQKLNSNTQV","FVLDEADEMLSRGFKNQIYDIFQKLNSNTQV","FVLDEADEMLSRGFKNQIYDIFQKLNSNTQV","MITTDLLARGIDVQQDSLVINYDLPTNRENY","INMVTEEDKRTLRDIETFYNTSIEENAPQCC","INMVTEEDKRTLRDIETFYNTSIEVNAPQCC"],[0,0,0,0,0,0,0,0,1,1,1,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],["chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr17:7577114_G/A","chr17:7577114_G/A","chr17:7577114_G/A","chr17:7577120_G/T","chr17:7577120_G/T","chr17:7577120_G/T","chr17:7577121_G/A","chr17:7577121_G/A","chr17:7577121_G/A","chr17:7578190_G/A","chr17:7578454_G/GA","chr17:7578455_T/TG"],["CDKN2A_A148E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A148E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A148E","EIF4A1_G192G","EIF4A1_G192G","EIF4A1_G192G","EIF4A1_K193N","EIF4A1_K193N","EIF4A1_K193N","EIF4A1_D194N","EIF4A1_D194N","EIF4A1_D194N","EIF4A1_V341D","EIF4A1_397fs","EIF4A1_397fs"],[93,93,93,93,93,93,93,93,90,90,90,93,93,93,93,93,93,93,93,93],[-21,0,-21,-24,0,-21,-21,-21,0,0,0,0,0,0,0,0,0,0,-27,-27]],"container":"<table class=\"compact stripe hover\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>seqnames<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>QUAL<\/th>\n      <th>sampleNames<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>cov_ref<\/th>\n      <th>cov_alt<\/th>\n      <th>varAllele<\/th>\n      <th>CDSLOC.start<\/th>\n      <th>CDSLOC.end<\/th>\n      <th>CDSLOC.width<\/th>\n      <th>PROTEINLOC<\/th>\n      <th>QUERYID<\/th>\n      <th>TXID<\/th>\n      <th>CDSID<\/th>\n      <th>GENEID<\/th>\n      <th>CONSEQUENCE<\/th>\n      <th>REFCODON<\/th>\n      <th>VARCODON<\/th>\n      <th>REFAA<\/th>\n      <th>VARAA<\/th>\n      <th>gene_name<\/th>\n      <th>tx_name<\/th>\n      <th>ref_nuc<\/th>\n      <th>ref_prot<\/th>\n      <th>alt_nuc<\/th>\n      <th>alt_prot<\/th>\n      <th>silent_start<\/th>\n      <th>silent_end<\/th>\n      <th>var_id<\/th>\n      <th>mut_id<\/th>\n      <th>alt_nnuc<\/th>\n      <th>alt_shift<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,6,10,11,13,14,15,17,32,33,36,37]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"seqnames","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"QUAL","targets":6},{"name":"sampleNames","targets":7},{"name":"ref","targets":8},{"name":"alt","targets":9},{"name":"cov_ref","targets":10},{"name":"cov_alt","targets":11},{"name":"varAllele","targets":12},{"name":"CDSLOC.start","targets":13},{"name":"CDSLOC.end","targets":14},{"name":"CDSLOC.width","targets":15},{"name":"PROTEINLOC","targets":16},{"name":"QUERYID","targets":17},{"name":"TXID","targets":18},{"name":"CDSID","targets":19},{"name":"GENEID","targets":20},{"name":"CONSEQUENCE","targets":21},{"name":"REFCODON","targets":22},{"name":"VARCODON","targets":23},{"name":"REFAA","targets":24},{"name":"VARAA","targets":25},{"name":"gene_name","targets":26},{"name":"tx_name","targets":27},{"name":"ref_nuc","targets":28},{"name":"ref_prot","targets":29},{"name":"alt_nuc","targets":30},{"name":"alt_prot","targets":31},{"name":"silent_start","targets":32},{"name":"silent_end","targets":33},{"name":"var_id","targets":34},{"name":"mut_id","targets":35},{"name":"alt_nnuc","targets":36},{"name":"alt_shift","targets":37}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><!--
### Phasing with germline variants


``` r
# not implemented yet
```
-->
</div>
</div>
<div class="section level2">
<h2 id="fusion-genes-from-rna-seq">Fusion genes from RNA-seq<a class="anchor" aria-label="anchor" href="#fusion-genes-from-rna-seq"></a>
</h2>
<div class="section level3">
<h3 id="reading-a-fusion-vcf">Reading a fusion VCF<a class="anchor" aria-label="anchor" href="#reading-a-fusion-vcf"></a>
</h3>
<p>First we want to read the fusion genes from a combined
<code>vcf</code> file like the one produced by the <a href="https://nf-co.re/rnafusion/" class="external-link"><code>rnafusion</code> NF-core
pipeline</a>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vr2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">readVcfAsVRanges</a></span><span class="op">(</span><span class="va">fusion_vcf_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_fusions.html">filter_fusions</a></span><span class="op">(</span>min_reads<span class="op">=</span><span class="fl">2</span>, min_split_reads<span class="op">=</span><span class="fl">1</span>, min_tools<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">vr2</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span></code></pre></div>
<p>Here, we are using the following filters:</p>
<ul>
<li>
<code>min_reads=2</code> – the fusion needs to be supported by 2
split or pair distance reads</li>
<li>
<code>min_split_reads=1</code> – the fusion needs to be supported by
at least one split read</li>
<li>
<code>min_tools=1</code> – the fusion needs to be reported by at
least one tool</li>
</ul>
</div>
<div class="section level3">
<h3 id="annotating-fusion-genes">Annotating fusion genes<a class="anchor" aria-label="anchor" href="#annotating-fusion-genes"></a>
</h3>
<p>Next, we subset the peptide context analogous to the SNPs:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fus</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_fusions.html">annotate_fusions</a></span><span class="op">(</span><span class="va">vr2</span>, <span class="va">ens106</span>, <span class="va">asm</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/subset_context_fusion.html">subset_context_fusion</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p>The <code>fus</code> table then looks like the following:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"data":[["1"],["NA--NA.fs"],[2],[1],[null],["chr22"],["+"],[23289621],["ENST00000305877"],["ENSG00000186716"],["AAACTCCAGACTGTCCACAGCATTCCGCTGACCATCAATAAGGAAGATGATGAGTCTCCGGGGCTCTATGGGTTTCTGAATGTCATCGTCCAC"],[2707],["chr9"],["+"],[130835447],["ENST00000318560"],["ENSG00000097007"],["ATGTTGGAGATCTGCCTGAAGCTGGTGGGCTGCAAATCCAAGAAGGGGCTGTCCTCGTCCTCCAGCTGTTATCTGGAAGAAGCCCTTCAGCGG"],[1],["AAACTCCAGACTGTCCACAGCATTCCGCTGACCATCAATAAGGAAGATGTTGGAGATCTGCCTGAAGCTGGTGGGCTGCAAATCCAAGAAGGGGCTGTCCTCGTCCTCCAGCTGTTATCTGGAAGAAGCCCTTCAGCGGCCAGTAGCATC"],[0]],"container":"<table class=\"compact stripe hover\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>fusion<\/th>\n      <th>split_reads<\/th>\n      <th>split_pairs<\/th>\n      <th>FFPM<\/th>\n      <th>seq_name_5p<\/th>\n      <th>seq_strand_5p<\/th>\n      <th>seq_start_5p<\/th>\n      <th>tx_id_5p<\/th>\n      <th>gene_id_5p<\/th>\n      <th>ref_nuc_5p<\/th>\n      <th>break_cdsloc_5p<\/th>\n      <th>seq_name_3p<\/th>\n      <th>seq_strand_3p<\/th>\n      <th>seq_start_3p<\/th>\n      <th>tx_id_3p<\/th>\n      <th>gene_id_3p<\/th>\n      <th>ref_nuc_3p<\/th>\n      <th>break_cdsloc_3p<\/th>\n      <th>alt_nuc<\/th>\n      <th>alt_shift<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,7,11,14,18,20]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"fusion","targets":1},{"name":"split_reads","targets":2},{"name":"split_pairs","targets":3},{"name":"FFPM","targets":4},{"name":"seq_name_5p","targets":5},{"name":"seq_strand_5p","targets":6},{"name":"seq_start_5p","targets":7},{"name":"tx_id_5p","targets":8},{"name":"gene_id_5p","targets":9},{"name":"ref_nuc_5p","targets":10},{"name":"break_cdsloc_5p","targets":11},{"name":"seq_name_3p","targets":12},{"name":"seq_strand_3p","targets":13},{"name":"seq_start_3p","targets":14},{"name":"tx_id_3p","targets":15},{"name":"gene_id_3p","targets":16},{"name":"ref_nuc_3p","targets":17},{"name":"break_cdsloc_3p","targets":18},{"name":"alt_nuc","targets":19},{"name":"alt_shift","targets":20}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="section level2">
<h2 id="generating-the-minigene-library">Generating the Minigene Library<a class="anchor" aria-label="anchor" href="#generating-the-minigene-library"></a>
</h2>
<div class="section level3">
<h3 id="tiling-cdnas-of-interest-into-smaller-peptides">Tiling cDNAs of interest into smaller peptides<a class="anchor" aria-label="anchor" href="#tiling-cdnas-of-interest-into-smaller-peptides"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tiled</span> <span class="op">=</span> <span class="fu"><a href="../reference/pep_tile.html">pep_tile</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/remove_cutsite.html">remove_cutsite</a></span><span class="op">(</span>BbsI<span class="op">=</span><span class="st">"GAAGAC"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>tiled</code> peptide table looks like the following:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22"],["chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr9:21970916_G/T","chr17:7577114_G/A","chr17:7577114_G/A","chr17:7577120_G/T","chr17:7577120_G/T","chr17:7577121_G/A","chr17:7577121_G/A","chr17:7578190_G/A","chr17:7578190_G/A","chr17:7578454_G/GA","chr17:7578454_G/GA","chr17:7578454_G/GA","chr17:7578455_T/TG"],["CDKN2A_A148E","CDKN2A_A148E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A97E","CDKN2A_A148E","CDKN2A_A148E","EIF4A1_G192G","EIF4A1_G192G","EIF4A1_K193N","EIF4A1_K193N","EIF4A1_D194N","EIF4A1_D194N","EIF4A1_V341D","EIF4A1_V341D","EIF4A1_397fs","EIF4A1_397fs","EIF4A1_397fs","EIF4A1_397fs"],["CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","CDKN2A","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1","EIF4A1"],["ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000147889","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960","ENSG00000161960"],["ENST00000304494","ENST00000304494","ENST00000479692","ENST00000479692","ENST00000494262","ENST00000494262","ENST00000497750","ENST00000497750","ENST00000498124","ENST00000498124","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831","ENST00000293831"],["ref","alt","ref","alt","ref","alt","ref","alt","ref","alt","ref","alt","ref","alt","ref","alt","ref","alt","ref","ref","alt","alt"],["GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGGTGAGGAC","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGGTGAGGAC","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAATGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGAACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGTTTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGATTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCTGACCTCATC","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCTGACCTCATC","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAAATGCCCCTCAATGTTGC","ATTAACATGGTGACAGAAGAAGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGTGAATGCCCCTCAATGTTGC"],["CDKN2A_A148-1","CDKN2A_A148E-1","CDKN2A_A97-1","CDKN2A_A97E-1","CDKN2A_A97-2","CDKN2A_A97E-2","CDKN2A_A97-3","CDKN2A_A97E-3","CDKN2A_A148-2","CDKN2A_A148E-2","EIF4A1_G192","EIF4A1_G192G","EIF4A1_K193","EIF4A1_K193N","EIF4A1_D194","EIF4A1_D194N","EIF4A1_V341","EIF4A1_V341D","EIF4A1_397-1","EIF4A1_397-2","EIF4A1_397fs-1","EIF4A1_397fs-2"],["GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGTTACAGCTAGCATCCAGGTCCCGGGAGGTGAA","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGACATCCCCGAT","GTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGACATCCCCGAT","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGGTGAGGAC","GATGTCGCACGGTACCTGCGCGCGGCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGGTGAGGAC","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGCGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","GCTGCGGGGGGCACCAGAGGCAGTAACCATGCCCGCATAGATGCCGAGGAAGGTCCCTCAGAAATGATCGGAAACCATTTGTGGGTTTGTAGA","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","AAGATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAAC","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","ATGTTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAATGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAG","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGGACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTTGTACTGGATGAAGCTGACGAAATGTTAAGCCGTGGATTCAAGAACCAGATCTATGACATATTCCAAAAGCTCAACAGCAACACCCAGGTA","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGTTTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","TTGATTACCACTGACCTGCTGGCCAGAGGCATTGATGTGCAGCAGGATTCTTTAGTCATCAACTATGACCTTCCCACCAACAGGGAAAACTAT","ATTAACATGGTGACAGAAGAGGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCT","GTGACAGAAGAGGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAATGCCCCTCAATGTTGCTGACCTCATC","ATTAACATGGTGACAGAAGAGGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGAAAATGCCCCTCAATGTTGC","ATTAACATGGTGACAGAAGAGGACAAGAGGACTCTTCGAGACATTGAGACCTTCTACAACACCTCCATTGAGGTGAATGCCCCTCAATGTTGC"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,1,1],[93,93,93,93,93,93,93,93,93,93,90,90,93,93,93,93,93,93,93,93,93,93],["VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","AAGGTRGSNHARIDAAEGPSVTASIQVPGGE","AAGGTRGSNHARIDAEEGPSVTASIQVPGGE","VARYLRAAAGGTRGSNHARIDAAEGPSDIPD","VARYLRAAAGGTRGSNHARIDAEEGPSDIPD","DVARYLRAAAGGTRGSNHARIDAAEGPSGED","DVARYLRAAAGGTRGSNHARIDAEEGPSGED","AAGGTRGSNHARIDAAEGPSEMIGNHLWVCR","AAGGTRGSNHARIDAEEGPSEMIGNHLWVCR","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","KMFVLDEADEMLSRGFKDQIYDIFQKLNSN","MFVLDEADEMLSRGFKDQIYDIFQKLNSNTQ","MFVLDEADEMLSRGFNDQIYDIFQKLNSNTQ","FVLDEADEMLSRGFKDQIYDIFQKLNSNTQV","FVLDEADEMLSRGFKNQIYDIFQKLNSNTQV","LITTDLLARGIDVQQVSLVINYDLPTNRENY","LITTDLLARGIDVQQDSLVINYDLPTNRENY","INMVTEEDKRTLRDIETFYNTSIEEMPLNVA","VTEEDKRTLRDIETFYNTSIEEMPLNVADLI","INMVTEEDKRTLRDIETFYNTSIEENAPQCC","INMVTEEDKRTLRDIETFYNTSIEVNAPQCC"],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1]],"container":"<table class=\"compact stripe hover\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>var_id<\/th>\n      <th>mut_id<\/th>\n      <th>gene_name<\/th>\n      <th>gene_id<\/th>\n      <th>tx_id<\/th>\n      <th>type<\/th>\n      <th>cDNA<\/th>\n      <th>pep_id<\/th>\n      <th>tiled<\/th>\n      <th>n_tiles<\/th>\n      <th>nt<\/th>\n      <th>peptide<\/th>\n      <th>BbsI_replaced<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[10,11,13]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"var_id","targets":1},{"name":"mut_id","targets":2},{"name":"gene_name","targets":3},{"name":"gene_id","targets":4},{"name":"tx_id","targets":5},{"name":"type","targets":6},{"name":"cDNA","targets":7},{"name":"pep_id","targets":8},{"name":"tiled","targets":9},{"name":"n_tiles","targets":10},{"name":"nt","targets":11},{"name":"peptide","targets":12},{"name":"BbsI_replaced","targets":13}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><!--
### Adding barcodes


``` r
# this is not yet implemented
```

__ SHOW TABLE HERE __
-->
</div>
<div class="section level3">
<h3 id="saving-the-report-file">Saving the report file<a class="anchor" aria-label="anchor" href="#saving-the-report-file"></a>
</h3>
<p>We can then combine our generated tables into a report save it with
the <code>writexl</code> package. We can include all tables listed
above:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report</span> <span class="op">=</span> <span class="fu"><a href="../reference/make_report.html">make_report</a></span><span class="op">(</span><span class="va">ann</span>, <span class="va">subs</span>, <span class="va">fus</span>, <span class="va">tiled</span><span class="op">)</span></span>
<span><span class="fu">writexl</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/writexl//reference/write_xlsx.html" class="external-link">write_xlsx</a></span><span class="op">(</span><span class="va">report</span>, <span class="st">"report_file.xlsx"</span><span class="op">)</span></span></code></pre></div>
<p>This <code>.xlsx</code> file will contain the different tables as
sheets. We will use it as an annotation file in the quality control and
screen steps.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Schubert.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
